File: templates/Uwp/_composition/_shared/Page.Ink.AddBasicInkServices/Services/Ink/InkFileService.cs
using Microsoft.Graphics.Canvas;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Windows.Foundation;
using Windows.Storage;
using Windows.Storage.Pickers;
using Windows.Storage.Provider;
using Windows.UI.Xaml.Controls;

namespace Param_ItemNamespace.Services.Ink
{
    public class InkFileService
    {
        private readonly InkCanvas _inkCanvas;
        private readonly InkStrokesService _strokesService;

        public InkFileService(InkCanvas inkCanvas, InkStrokesService strokesService)
        {
            _inkCanvas = inkCanvas;
            _strokesService = strokesService;
        }

        public async Task<bool> LoadInkAsync()
        {
            var openPicker = new FileOpenPicker
            {
                SuggestedStartLocation = PickerLocationId.PicturesLibrary
            };

            openPicker.FileTypeFilter.Add(".gif");

            var file = await openPicker.PickSingleFileAsync();
            return await _strokesService.LoadInkFileAsync(file);
        }

        public async Task SaveInkAsync()
        {
            if (!_strokesService.GetStrokes().Any())
            {
                return;
            }

            var savePicker = new FileSavePicker
            {
                SuggestedStartLocation = PickerLocationId.PicturesLibrary
            };

            savePicker.FileTypeChoices.Add("Gif with embedded ISF", new List<string> { ".gif" });

            var file = await savePicker.PickSaveFileAsync();
            await _strokesService.SaveInkFileAsync(file);
        }

        public async Task<StorageFile> ExportToImageAsync(StorageFile imageFile = null)
        {
            if (!_strokesService.GetStrokes().Any())
            {
                return null;
            }

            if (imageFile != null)
            {
                return await ExportCanvasAndImageAsync(imageFile);
            }
            else
            {
                return await ExportCanvasAsync();
            }
        }

        private async Task<StorageFile> ExportCanvasAndImageAsync(StorageFile imageFile)
        {
            var saveFile = await GetImageToSaveAsync();

            if (saveFile == null)
            {
                return null;
            }

            // Prevent updates to the file until updates are finalized with call to CompleteUpdatesAsync.
            CachedFileManager.DeferUpdates(saveFile);

            using (var outStream = await saveFile.OpenAsync(FileAccessMode.ReadWrite))
            {
                var device = CanvasDevice.GetSharedDevice();

                CanvasBitmap canvasbitmap;
                using (var stream = await imageFile.OpenAsync(FileAccessMode.Read))
                {
                    canvasbitmap = await CanvasBitmap.LoadAsync(device, stream);
                }

                using (var renderTarget = new CanvasRenderTarget(device, (int)_inkCanvas.Width, (int)_inkCanvas.Height, canvasbitmap.Dpi))
                {
                    using (CanvasDrawingSession ds = renderTarget.CreateDrawingSession())
                    {
                        ds.DrawImage(canvasbitmap, new Rect(0, 0, (int)_inkCanvas.Width, (int)_inkCanvas.Height));
                        ds.DrawInk(_strokesService.GetStrokes());
                    }

                    await renderTarget.SaveAsync(outStream, CanvasBitmapFileFormat.Png);
                }
            }

            // Finalize write so other apps can update file.
            FileUpdateStatus status = await CachedFileManager.CompleteUpdatesAsync(saveFile);

            return saveFile;
        }

        private async Task<StorageFile> ExportCanvasAsync()
        {
            var file = await GetImageToSaveAsync();

            if (file == null)
            {
                return null;
            }

            CanvasDevice device = CanvasDevice.GetSharedDevice();
            CanvasRenderTarget renderTarget = new CanvasRenderTarget(device, (int)_inkCanvas.Width, (int)_inkCanvas.Height, 96);

            using (var ds = renderTarget.CreateDrawingSession())
            {
                ds.DrawInk(_strokesService.GetStrokes());
            }

            using (var fileStream = await file.OpenAsync(FileAccessMode.ReadWrite))
            {
                await renderTarget.SaveAsync(fileStream, CanvasBitmapFileFormat.Png);
            }

            return file;
        }

        private async Task<StorageFile> GetImageToSaveAsync()
        {
            var savePicker = new FileSavePicker();
            savePicker.SuggestedStartLocation = PickerLocationId.PicturesLibrary;
            savePicker.FileTypeChoices.Add("PNG", new List<string>() { ".png" });
            var saveFile = await savePicker.PickSaveFileAsync();

            return saveFile;
        }
    }
}

File: templates/Uwp/_comp/_shared/Page.Ink.AddBasicInkServices._VB/Services/Ink/InkFileService.vb
Imports Microsoft.Graphics.Canvas
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Threading.Tasks
Imports Windows.Foundation
Imports Windows.Storage
Imports Windows.Storage.Pickers
Imports Windows.Storage.Provider

Namespace Services.Ink
    Public Class InkFileService
        Private ReadOnly _inkCanvas As InkCanvas
        Private ReadOnly _strokesService As InkStrokesService

        Public Sub New(inkCanvas As InkCanvas, strokesService As InkStrokesService)
            _inkCanvas = inkCanvas
            _strokesService = strokesService
        End Sub

        Public Async Function LoadInkAsync() As Task(Of Boolean)
            Dim openPicker = New FileOpenPicker With {
                .SuggestedStartLocation = PickerLocationId.PicturesLibrary
            }
            openPicker.FileTypeFilter.Add(".gif")
            Dim file = Await openPicker.PickSingleFileAsync()
            Return Await _strokesService.LoadInkFileAsync(file)
        End Function

        Public Async Function SaveInkAsync() As Task
            If Not _strokesService.GetStrokes().Any() Then
                Return
            End If

            Dim savePicker = New FileSavePicker With {
                .SuggestedStartLocation = PickerLocationId.PicturesLibrary
            }
            savePicker.FileTypeChoices.Add("Gif with embedded ISF", New List(Of String) From {
                ".gif"
            })
            Dim file = Await savePicker.PickSaveFileAsync()
            Await _strokesService.SaveInkFileAsync(file)
        End Function

        Public Async Function ExportToImageAsync(Optional imageFile As StorageFile = Nothing) As Task(Of StorageFile)
            If Not _strokesService.GetStrokes().Any() Then
                Return Nothing
            End If

            If imageFile IsNot Nothing Then
                Return Await ExportCanvasAndImageAsync(imageFile)
            Else
                Return Await ExportCanvasAsync()
            End If
        End Function

        Private Async Function ExportCanvasAndImageAsync(imageFile As StorageFile) As Task(Of StorageFile)
            Dim saveFile = Await GetImageToSaveAsync()

            If saveFile Is Nothing Then
                Return Nothing
            End If

            ' Prevent updates to the file until updates are finalized with call to CompleteUpdatesAsync.
            CachedFileManager.DeferUpdates(saveFile)

            Using outStream = Await saveFile.OpenAsync(FileAccessMode.ReadWrite)
                Dim device = CanvasDevice.GetSharedDevice()
                Dim canvasbitmap As CanvasBitmap

                Using stream = Await imageFile.OpenAsync(FileAccessMode.Read)
                    canvasbitmap = Await CanvasBitmap.LoadAsync(device, stream)
                End Using

                Using renderTarget = New CanvasRenderTarget(device, CInt(_inkCanvas.Width), CInt(_inkCanvas.Height), canvasbitmap.Dpi)

                    Using ds As CanvasDrawingSession = renderTarget.CreateDrawingSession()
                        ds.DrawImage(canvasbitmap, New Rect(0, 0, CInt(_inkCanvas.Width), CInt(_inkCanvas.Height)))
                        ds.DrawInk(_strokesService.GetStrokes())
                    End Using

                    Await renderTarget.SaveAsync(outStream, CanvasBitmapFileFormat.Png)
                End Using
            End Using

            '  Finalize write so other apps can update file.
            Await CachedFileManager.CompleteUpdatesAsync(saveFile)
            Return saveFile
        End Function

        Private Async Function ExportCanvasAsync() As Task(Of StorageFile)
            Dim file = Await GetImageToSaveAsync()

            If file Is Nothing Then
                Return Nothing
            End If

            Dim device As CanvasDevice = CanvasDevice.GetSharedDevice()
            Dim renderTarget As CanvasRenderTarget = New CanvasRenderTarget(device, CInt(_inkCanvas.Width), CInt(_inkCanvas.Height), 96)

            Using ds = renderTarget.CreateDrawingSession()
                ds.DrawInk(_strokesService.GetStrokes())
            End Using

            Using fileStream = Await file.OpenAsync(FileAccessMode.ReadWrite)
                Await renderTarget.SaveAsync(fileStream, CanvasBitmapFileFormat.Png)
            End Using

            Return file
        End Function

        Private Async Function GetImageToSaveAsync() As Task(Of StorageFile)
            Dim savePicker = New FileSavePicker()
            savePicker.SuggestedStartLocation = PickerLocationId.PicturesLibrary
            savePicker.FileTypeChoices.Add("PNG", New List(Of String)() From {
                ".png"
            })
            Dim saveFile = Await savePicker.PickSaveFileAsync()
            Return saveFile
        End Function
    End Class
End Namespace
